src/app/components/appointmentform/actions.ts

```diff
@@ -0,0 +1,93 @@
+"use server";
+
+import {
+	Appointment,
+	createAppointment,
+	getAppointmentsBySchedule,
+	getAppointmentByID,
+} from "@/db/homeseeker/appointment";
+import { getScheduleByID } from "@/db/homeseeker/schedule";
+import { getCurrentUser } from "@/app/cookies";
+
+export async function makeAppointment(
+	schedule_id: number,
+	start_time: Date,
+	end_time: Date,
+): Promise<Appointment | string> {
+	// Errors checking
+	if (!start_time) {
+		return "Start time required.";
+	}
+	if (!end_time) {
+		return "End time required.";
+	}
+
+	// Check if the start time and end time are valid
+	if (start_time > end_time) {
+		return "Start time must be earlier than end time.";
+	}
+
+	// Get user currently logged in
+	const user = getCurrentUser();
+	if (!user) {
+		return "Not signed in";
+	}
+
+	// Check if start and end time are inside scheduled times
+	const schedule = await getScheduleByID(schedule_id);
+	if (!schedule) {
+		return "Error retreiving schedule";
+	}
+	if (schedule.start > start_time) {
+		return "Start time must be within scheduled times."
+	}
+	if (schedule.end < end_time) {
+		return "End time must be within scheduled times."
+	}
+
+	// Make sure the duration is between 20 and 120 mintues
+	const duration = (end_time.getTime() - start_time.getTime()) / 60000;
+	if (duration < 20) {
+		return "Appointments must be at least 20 minutes long.";
+	}
+	if (duration > 180) {
+		return "Appointments must be less than 3 hours long.";
+	}
+
+	// Checks for overlaps with
+	const appointments = await getAppointmentsBySchedule(schedule_id);
+	for (const appointment of appointments) {
+		if (start_time < appointment.start) {
+			if (end_time > appointment.start) {
+				return "Time conflicts with another appointment.";
+			}
+		} else {
+			if (start_time < appointment.end) {
+				return "Time conflicts with another appointment.";
+			}
+		}
+	}
```

Comment:
We could probably convert this to a SQL query which would be faster, but I'm fine with just leaving it as-is for the time being.

Reply:
Yea probably faster with SQL.